### DISCUSSION TITLE
Improve ListObjects max-out behavior

### DISCUSSION THREAD

**Discussion by nediamond** [openfga/community#304] [Ideas]
In the current state, there is no direct signal as to whether the result of [ListObjects](https://openfga.dev/api/service#/Relationship%20Queries/ListObjects) is complete or not, ie whether or not `OPENFGA_LIST_OBJECTS_DEADLINE` or `OPENFGA_LIST_OBJECTS_MAX_RESULTS` are hit. The latter could potentially be inferred by callers, but not actually with full-confidence (incorrect if real n=limit). This information gap could result in emergence of unsafe usage by assuming completeness of a response when it is not.

I believe the ideal solution for this would be to add pagination to the `ListObjects` api via a `continuation_token` pattern that is consistent with how it is done for [Tuples/Read](https://openfga.dev/api/service#/Relationship%20Tuples/Read), though I am not sure if there are substantial technical blockers to that which I am unaware of.

A minimally viable solution could be something like an additional property on the `ListObjects` response `incomplete_reason` which can be `null, 'OPENFGA_LIST_OBJECTS_DEADLINE', 'OPENFGA_LIST_OBJECTS_MAX_RESULTS'` or a functional equivalent of that. Not sure what a technical blocker to that could be.

Design questions raised here may also be relevant to the upcoming `ListUsers` API.

**Comment by evansims** [comment]
Hey @nediamond 👋 Thanks for raising this! The team has recently been talking about this internally as well. We've been thinking through some different approaches to improving that experience. We really appreciate your feedback and will definitely add it to the conversation. We'll have more to share on this later.

**Comment by willvedd** [comment]
@nediamond You raise a good point. In practice, are you finding that you can't determine if the max results are hit or is this a matter of confidence? We'd like to understand your use case a bit more.

Technically speaking, the `continuation_token` pattern you suggested is unfortunately a non-starter for ListObjects and ListUsers provided how their algorithms work. However, an additional "reason" property in the response that indicates if the request timed-out or hit max results is feasible. Though no promises on delivery; we'll need to think through it a little more.

Thank you for the thoughtful suggestion!

**Comment by nazar-kostiv-moodys** [comment]
Good morning. I don't believe the functionality described in this thread has been added to OpenFGA as of v1.8.16. Are there any plans to do so or were some technical limitations hit to prevent it?

For reference our current solution is to run a parallel timeout task in our codebase alongside openFGA listObjects call. But set our timeout to be slightly lower than what we set for `OPENFGA_LIST_OBJECTS_DEADLINE`. Then if timeout task completes first, we error out on our end.

This however is rather fiddly and not a good coding practise overall as it pulls config into the code.
For our purposes, both erroring out if deadline is reached or an indication of any sort would be sufficient.

Any updates on this would be greatly appreciated.

**Reply by dyeam0** [reply]
Hello @nazar-kostiv-moodys, that functionality has not been added. It is not in our plan currently, so we cannot give you a timeline.

You may want to try the [/streamed-list-objects](https://openfga.dev/api/service#/Relationship%20Queries/StreamedListObjects) endpoint. A different approach where results are streamed to the client. It will stream up to 3 seconds before timing out.
